### PR TITLE
Use Github Actions to build FreeOrion

### PR DESCRIPTION
**Pull request by o01eg**

### CODE REVIEW

**Comment by geoffthemedio** [comment]
Does using GitHub actions require not using AppVeyor?

**Comment by o01eg** [comment]
@geoffthemedio Not, it just moves Windows CI from Appveyor to Github Actions.

**Comment by geoffthemedio** [comment]
> ... moves Windows CI from Appveyor to Github Actions.

Why do you want to do that?

**Comment by o01eg** [comment]
I remember we had a plan to move from Appveyor like FreeOrion SDK did.

File: .github/workflows/build.yml
Comment (geoffthemedio) [inline .github/workflows/build.yml:11]: does this mean it will only rub test builds on the master branch when it is pushed an no others unless they are part of a pull request?

Comment (o01eg) [reply .github/workflows/build.yml:11]: Yes. Do we need to build pushes always?

Comment (geoffthemedio) [reply .github/workflows/build.yml:11]: I find it handy to have a test branch that I can push to test build without needing a pull request always open. Could just add the [refactor_test](https://github.com/freeorion/freeorion/tree/refactor_test) branch for that, though, I suppose.

**Review by Vezzra — COMMENTED** [review]
I'm sorry, but I don't think I can provide any meaningful input here - my knowledge about all the CI stuff is next to non-existent.

For what it's worth, looks fine from what I can see.

File: .github/workflows/build.yml
Comment (Cjkjvfnby) [reply .github/workflows/build.yml:11]: We can add a special pattern for that.  For example `ci_*`

https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#onpushpull_requestbranchestags

File: .github/workflows/mono-uninstall.sh
Comment (Cjkjvfnby) [inline .github/workflows/mono-uninstall.sh:3]: Could you share more info on why do we need this step?

**Review by Cjkjvfnby — COMMENTED** [review]
I don't understand this line in logs.  Is it expected?

image

**Comment by Cjkjvfnby** [comment]
The game builds on Ubuntu in `25m 11s`  is it normal numbers?

**Comment by geoffthemedio** [comment]
> The game builds on Ubuntu in `25m 11s` is it normal numbers?

yes, eg: https://travis-ci.org/github/freeorion/freeorion/builds/753497352

**Comment by Cjkjvfnby** [comment]
> yes, eg: https://travis-ci.org/github/freeorion/freeorion/builds/753497352

What is the timing on your machine?

**Comment by geoffthemedio** [comment]
> What is the timing on your machine?

About 4 min with parallel builds on a 12-core desktop...
````
1>------ Build started: Project: Parsers, Configuration: Release Win32 ------
2>------ Build started: Project: GiGi, Configuration: Release Win32 ------
1>Build started 2021-01-10 12:30:58 PM.
2>Build started 2021-01-10 12:30:58 PM.
2>Time Elapsed 00:00:30.86
1>Time Elapsed 00:01:39.36
3>------ Build started: Project: Common, Configuration: Release Win32 ------
3>Build started 2021-01-10 12:32:37 PM.
3>Time Elapsed 00:00:39.40
4>------ Build started: Project: FreeOrion, Configuration: Release Win32 ------
5>------ Build started: Project: FreeOrionD, Configuration: Release Win32 ------
6>------ Build started: Project: FreeOrionCA, Configuration: Release Win32 ------
5>Build started 2021-01-10 12:33:17 PM.
4>Build started 2021-01-10 12:33:17 PM.
6>Build started 2021-01-10 12:33:17 PM.
4>Time Elapsed 00:01:03.35
6>Time Elapsed 00:01:13.77
5>Time Elapsed 00:01:24.65
========== Build: 6 succeeded, 0 failed, 0 up-to-date, 0 skipped ==========
````
My older 4-core laptop took 30-40 min or so, I think.

**Comment by Cjkjvfnby** [comment]
Also, we need to add paths that should trigger build.  There is no sense to run compilation if the translation is changed.

We should cover cpp and build scrips(cmake, github actions, etc)

Since build takes a lot of time this thing worth investment.

https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#onpushpull_requestpaths

**Comment by o01eg** [comment]
We could cover content as well because it could be tested with unit tests.

File: .github/workflows/build.yml
Comment (geoffthemedio) [reply .github/workflows/build.yml:70]: Whether the human client is built, mainly, I think?

**Comment by Cjkjvfnby** [comment]
> We could cover content as well because it could be tested with unit tests.

`unit tests` is a very confusing term, it is the name of the library that used to do unit, component/integration, and end-to-end tests.  If we are talking about real unit tests most of them are checking logic and can be run only on one system (don't need to run them on all builds).

**Comment by Cjkjvfnby** [comment]
I have remembered that path filtering doesn't work with protection: https://docs.github.com/en/free-pro-team@latest/github/administering-a-repository/enabling-required-status-checks

File: .github/workflows/build.yml
Comment (Cjkjvfnby) [inline .github/workflows/build.yml:41]: The timestamp for a cache looks strange to me. I don't understand how it could be reused.

I am not sure what is really cached, so maybe I will tell you some nonsense. 

In general, I expect 2 cache levels. 

1) Dependencies. Key is md5 of file/files with dependencies.  This thing is not changed often, so it could be reused between builds.  
2) Freeorion generated cache.  We can use a commit as a key (this will allow us to reuse cache during actions run on the same commit).  A more elegant way is to use the last commit that touches C++/build files. In that case, we could reuse cache for PRs that change only the content of `default` folder.  The best way is to cache modules separately.

Comment (o01eg) [reply .github/workflows/build.yml:41]: As I understand it uses two caches with different action version to cache dependencies and ccache. And `restore-key` allows to use ccache with different timestamp.

Comment (o01eg) [reply .github/workflows/build.yml:41]: It's based on https://github.com/freeorion/freeorion/pull/3270 and https://github.com/airvzxf/cache-anything-new-action/issues/1

**Comment by o01eg** [comment]
I moved Linux build only because I'm not sure what broken on MacOS and Windows.

File: .github/workflows/mono-uninstall.sh
Comment (Cjkjvfnby) [inline .github/workflows/mono-uninstall.sh:1]: This file is not part of linux build, could you remove it from this PR?  Probably you need to save it for later.

File: .github/workflows/build.yml
Comment (Cjkjvfnby) [inline .github/workflows/build.yml]: I'd recommend rewording it to clarify that `ccached` is not a typo but a tool name.

`Cache files with ccache`

File: .github/workflows/build.yml
Comment (Cjkjvfnby) [inline .github/workflows/build.yml]: Let's use the full name for arguments in that script. This increases readability for people who is not familiar with that tool.   
`--show-stats`

File: .github/workflows/build.yml
Comment (Cjkjvfnby) [inline .github/workflows/build.yml]: We have 2 stats call. They are run in different directories, does it affects anything?

Could we move them to separate steps before and after the build step?  There might be many reasons to see the build log and checking cache is only one of them.  In most cases, cache statistics are not interesting for developers.  Also, you could remove echo parts since they will be in different sections.  Moreover, it would be much easier to compare them because build logs are huge.

Steps in UI:
image

File: .github/workflows/install-build.sh
Comment (Cjkjvfnby) [reply .github/workflows/install-build.sh]: @o01eg , could you rename this file?

File: .github/workflows/install-build.sh
Comment (Cjkjvfnby) [inline .github/workflows/install-build.sh]: Is it a part of the Linux build?  We have a separate build for linters, probably we don't need this here.

**Review by Cjkjvfnby — COMMENTED** [review]
My comments do not block the merge, most of them could be done later.

@geoffthemedio ,  someone should set build as required for PRs after merging. Could you check if you can do it?  See Require status checks to pass before merging in documentation https://docs.github.com/en/github/administering-a-repository/managing-a-branch-protection-rule

File: .github/workflows/mono-uninstall.sh
Comment (o01eg) [reply .github/workflows/mono-uninstall.sh:1]: It's a part of MacOS build.

File: .github/workflows/build.yml
Comment (o01eg) [reply .github/workflows/build.yml]: OK, fixed.

File: .github/workflows/build.yml
Comment (o01eg) [reply .github/workflows/build.yml]: Fixed.

File: .github/workflows/build.yml
Comment (o01eg) [reply .github/workflows/build.yml]: Moved to separate step.

File: .github/workflows/install-build.sh
Comment (o01eg) [reply .github/workflows/install-build.sh]: Renamed to `linux-install-build.sh` and removed linter installation.

File: .github/workflows/build.yml
Comment (Cjkjvfnby) [inline .github/workflows/build.yml:69]: @o01eg , failure on the test step does not fail the build. Is it intended?  `33% tests passed, 2 tests failed out of 3`

File: .github/workflows/build.yml
Comment (Cjkjvfnby) [inline .github/workflows/build.yml]: Lets name it `Run C++ unit tests` to avoid confusion with Python unit tests.

File: .github/workflows/build.yml
Comment (o01eg) [reply .github/workflows/build.yml:69]: Yes, it is intended. We get some strange failure here sometimes.

File: .github/workflows/build.yml
Comment (o01eg) [reply .github/workflows/build.yml]: Renamed.

File: .github/workflows/build.yml
Comment (Cjkjvfnby) [reply .github/workflows/build.yml:69]: This not how tests supposed to work. If they do not fail it is the same as not having them, but longer.

Do you know if someone has a plan to fix it? 

Anyway, this is definitely out of the scope of this PR.

**Comment by Cjkjvfnby** [comment]
@o01eg , do we need to do a check that if compilation is failed, this action is failed too?

File: .github/workflows/build.yml
Comment (o01eg) [reply .github/workflows/build.yml:69]: I cann't reproduce it so maybe someone else could fix it.

**Comment by o01eg** [comment]
No, I disabled MacOS and Windows because broken compilation made action to fail.

**Comment by geoffthemedio** [comment]
> @geoffthemedio , someone should set build as required for PRs after merging. 

Do you mean before merging? From what I see, there are configurable rules that branches must meet to be merged, but that doesn't seem related to automatically starting a test build *after* merging...

> Could you check if you can do it? See Require status checks to pass before merging in documentation https://docs.github.com/en/github/administering-a-repository/managing-a-branch-protection-rule

I see:
branch_protection_rule

Not sure what you want, though.

**Comment by Cjkjvfnby** [comment]
@geoffthemedio , 

This checkbox should be checked after merging this PR, to make build required for all branches before merging. 
image

**Comment by geoffthemedio** [comment]
> This checkbox should be checked after merging this PR, to make build required for all branches before merging.

As long as it doesn't time out ~40% of the builds...

**Review by Cjkjvfnby — APPROVED** [review]

**Comment by Cjkjvfnby** [comment]
> As long as it doesn't time out ~40% of the builds...

Am I right that build failures are already a part of the process and this PR does not change how things merged now?

That means that if we merge these changes and something went wrong, it won't be a huge trouble, so we can fix it without a strict deadline.

**Comment by hifron** [comment]
One huge Github rules them all... ?

**Review by Vezzra — COMMENTED** [review]
I'm afraid I still don't know what meaningful input I could give here.

Can only repeat what I said before: looks ok to me, but that doesn't count for much given I don't really know much about this stuff.

**Comment by geoffthemedio** [comment]
Is there a way to tell CMake or the test build to use some extra include directories to fix [this](https://github.com/freeorion/freeorion/pull/3263/checks?check_run_id=1947523282#step:8:354)?
````
In file included from /usr/include/x86_64-linux-gnu/SDL2/_real_SDL_config.h:33:
/usr/include/x86_64-linux-gnu/SDL2/SDL_platform.h:179:10: fatal error: 'begin_code.h' file not found
#include "begin_code.h"
         ^~~~~~~~~~~~~~
````
As discussed [here](https://bugs.launchpad.net/ubuntu/+source/libsdl2-ttf/+bug/1872023)...

**Comment by o01eg** [comment]
Docs tell about CXXFLAGS and CFLAGS https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_FLAGS_INIT.html#variable:CMAKE_%3CLANG%3E_FLAGS_INIT

**Comment by geoffthemedio** [comment]
Tried that, doesn't seem to help: https://github.com/freeorion/freeorion/commit/4d567c178ee6504f0e0a98ffcc0ce0437b488280

**Comment by o01eg** [comment]
@geoffthemedio It works for me, should be `CXXFLAGS` not `CPPFLAGS`.

**Comment by geoffthemedio** [comment]
> @geoffthemedio It works for me, should be `CXXFLAGS` not `CPPFLAGS`.

Can you tell specifically which include directory is needed for Ubuntu 20.04?

**Comment by o01eg** [comment]
@geoffthemedio I used same from `.travis.yml`.

```
-I/usr/include/SDL2 -I/usr/include/x86_64-linux-gnu/SDL2/ -I/usr/include/x86_64-linux-gnu/
```

**Comment by geoffthemedio** [comment]
> @geoffthemedio I used same from `.travis.yml`.
> 
> ```
> -I/usr/include/SDL2 -I/usr/include/x86_64-linux-gnu/SDL2/ -I/usr/include/x86_64-linux-gnu/
> ```

My question was which of those is really needed?Probably not all are.

**Comment by o01eg** [comment]
I think I can only find out it with trial and error.

**Comment by geoffthemedio** [comment]
It would probably also be better if it could be fixed directly in CMakeLists.txt rather than in the CI config.

**Comment by o01eg** [comment]
@geoffthemedio As I understand it is clang issue which is supposed to be fixed later.

**Comment by geoffthemedio** [comment]
> @geoffthemedio As I understand it is clang issue which is supposed to be fixed later.

Does that change how it should be fixed now for FreeOrion?

**Comment by o01eg** [comment]
@geoffthemedio Yes, I think so.

**Comment by o01eg** [comment]
Superseded by #3411 , #3409 and #3408

**Pull request closed without merging**
